aptos-move/move-examples/bonding_curve_launchpad/sources/liquidity_pair.move

```diff
@@ -0,0 +1,277 @@
+    use swap::router;
+    use swap::liquidity_pool;
+    use swap::coin_wrapper;
+    use resource_account::resource_signer_holder;
+    use resource_account::bonding_curve_launchpad;
+	friend bonding_curve_launchpad;
+
+    const FA_DECIMALS: u8 = 8;
+    const INITIAL_VIRTUAL_APT_LIQUIDITY: u128 = 50_000_000_000;
+    const APT_LIQUIDITY_THRESHOLD: u128 = 600_000_000_000;
+
+    const EFA_PRIMARY_STORE_DOES_NOT_EXIST: u64 = 12;
+    const ELIQUIDITY_PAIR_EXISTS_ALREADY: u64 = 100;
+    const ELIQUIDITY_PAIR_DOES_NOT_EXIST: u64 = 101;
+    const ELIQUIDITY_PAIR_DISABLED: u64 = 102;
+    const ELIQUIDITY_PAIR_SWAP_AMOUNTOUT_INSIGNIFICANT: u64 = 111;
+
+    //---------------------------Events---------------------------
+    #[event]
+    struct LiquidityPairCreated has store, drop {
+        fa_object_metadata: Object<Metadata>,
+        initial_fa_reserves: u128,
+        initial_apt_reserves: u128,
+        k: u256
+    }
+    #[event]
+    struct LiquidityPairReservesUpdated has store, drop {
+        old_fa_reserves: u128,
+        old_apt_reserves: u128,
+        new_fa_reserves: u128,
+        new_apt_reserves: u128
+    }
+    #[event]
+    struct LiquidityPairSwap has store, drop {
+        is_fa_else_apt: bool,
+        gained: u128,
+        swapper_address: address
+    }
+    #[event]
+    struct LiquidityPairGraduated has store, drop {
+        fa_object_metadata: Object<Metadata>,
+        dex_address: address
+    }
+
+    //---------------------------Structs---------------------------
+    struct LiquidityPairSmartTable has key {
+        liquidity_pairs: SmartTable<Object<Metadata>, LiquidityPair>
+    }
+    struct LiquidityPair has store {
+        is_enabled: bool,
+        is_frozen: bool,
+        fa_reserves: u128,
+        apt_reserves: u128,
+        k_constant: u256
+    }
+
+    //---------------------------Init---------------------------
+    fun init_module(account: &signer) {
+        let liquidity_pair_table: LiquidityPairSmartTable = LiquidityPairSmartTable {
+            liquidity_pairs: smart_table::new()
+        };
+
+        move_to(account, liquidity_pair_table);
+    }
+
+    //---------------------------Liquidity Pair---------------------------
+    public(friend) fun register_liquidity_pair(transfer_ref: &fungible_asset::TransferRef, account: &signer, fa_object_metadata: Object<Metadata>, apt_initialPurchaseAmountIn: u64, fa_initialLiquidity: u128) acquires LiquidityPairSmartTable {
+        let liquidity_pair_smartTable = borrow_global_mut<LiquidityPairSmartTable>(@resource_account);
+        assert!(!smart_table::contains(&mut liquidity_pair_smartTable.liquidity_pairs, fa_object_metadata), ELIQUIDITY_PAIR_EXISTS_ALREADY);
+        //* Initial APT reserves are virtual liquidity, for less extreme initial swaps (avoiding early adopter's advantage, for fairness).
+        let k_constant: u256 = (fa_initialLiquidity as u256) * (INITIAL_VIRTUAL_APT_LIQUIDITY as u256);
```

Comment:
Why do we need to explicitly store k if this can be computed from the reserves? It also doesn't seem to be used anywhere else in this module